Changes to have can tests working.

MMaiero opened the pull request:
It provides also some improvements in order to have tests passing in OS X.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

MMaiero committed:
Changes to have can tests working on osx.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

amitjoy commented on kura/org.eclipse.kura.protocol.can.test/src/main/java/org/eclipse/kura/protocol/can/test/CanSocketTest.java:
```diff
@@ -117,6 +101,37 @@ public void updated(Map<String, Object> properties) {
+
+            @Override
+            public void run() {
+                if (CanSocketTest.this.m_canConnection != null && !IS_MAC) {
```
`CanSocketTest.this.m_canConnection `and `CanSocketTest.this.thread_done` must be volatile

MMaiero replied:
`CanSocketTest.this.thread_done` does not need to be a class variable. I've changed the code accordingly.

amitjoy commented on kura/org.eclipse.kura.protocol.can.test/src/main/java/org/eclipse/kura/protocol/can/test/CanSocketTest.java:
```diff
@@ -117,6 +101,37 @@ public void updated(Map<String, Object> properties) {
+
+    @Test
+    public void startCanTestThread() {
+        if (this.m_pollThread != null) {
```
This code to let the thread finish if there already exists one is as same as in the `deactivate()` method. You can extract this code to a private method.

MMaiero replied:
Agreed.

amitjoy reviewed (changes requested):
Some lurking threading issue and minor code refactoring

MMaiero committed:
The following changes were applied:

- Renamed variables to meet the new format.
- Converted thread_done from a class variable to a local variable.
- Extracted terminatePollThread method.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

amitjoy commented on kura/org.eclipse.kura.protocol.can.test/src/main/java/org/eclipse/kura/protocol/can/test/CanSocketTest.java:
```diff
@@ -108,42 +99,35 @@ public void setup() {
         CanMessage cm = null;
-        if (this.m_isMaster) {
-            if (this.m_orig >= 0) {
+        if (this.isMaster) {
```
`isMaster` is also a state variable shared among threads. It needs to be volatile as well. So is `this.orig`.

MMaiero replied:
I can agree with you about `this.interfaceName`, `this.canId ` and `this.isMaster` that are variables that, in theory, can be affected by multithreading.
The others, in my opinion, are effectively managed by a single thread and do not need to be volatile.

amitjoy replied:
`orig` should also be volatile as it determines a state on which thread works. `canConnection` should also be volatile. Let's suppose the `canConnection` is updated with new implementation at runtime but the thread already has acquired a reference to the old instance of `canConnection`, hence, the thread will not see the updated reference of `canConnection`.

MMaiero replied:
canConnection **is** volatile.
In any case, due to the DS definition, if a new implementation of canConnection is available, taking the place of the old, the deactivate method will be invoked and the thread will be killed.

amitjoy replied:
Actually in OSGi, it works a bit differently. During dynamic service replacement, the bind of the "new" service happens before the unbind of the "old" service. Hence, volatile is a best fit for such things.

amitjoy commented on kura/org.eclipse.kura.protocol.can.test/src/main/java/org/eclipse/kura/protocol/can/test/CanSocketTest.java:
```diff
@@ -108,42 +99,35 @@ public void setup() {
+            if (this.orig >= 0) {
                 try {
-                    testSendImpl(this.m_ifName, this.m_canId, this.m_orig);
+                    testSendImpl(this.interfaceName, this.canId, this.orig);
```
`interfaceName` and `canId` also need to be volatile.

amitjoy commented on kura/org.eclipse.kura.protocol.can.test/src/main/java/org/eclipse/kura/protocol/can/test/CanSocketTest.java:
```diff
@@ -152,13 +136,11 @@ public boolean doCanTest() {
-
-                cm = this.m_canConnection.receiveCanMessage(-1, 0x7FF);
-
+                cm = this.canConnection.receiveCanMessage(-1, 0x7FF);
```
`canConnection` too

cdealti replied:
@amitjoy I believe you are right on requiring volatile on all the fields that are modified in the ``updated`` method the way it's written now. However this ``updated`` method does not follow our tipical pattern of:

1. Stop all the threads
2. Modify internal state
3. Restart the threads (the run method will see the last modified stated)

But there is a more fundamental flaw. This OSGi (integration test) will not even run from Maven or Eclipse because it does not use the approach described in [1].

In my opinion the test will be ignored on the Mac (because of the assumption in the ``setup`` method) and will always succeed without testing anything because this.canConnection is always null in startCanTestThread.

@markoer what do you think?

[1] https://github.com/eclipse/kura/blob/develop/kura/org.eclipse.kura.core.test/src/main/java/org/eclipse/kura/core/test/ExampleTest.java#L41

amitjoy replied:
You are right. It would not run successfully to inject the dependencies as in the OSGi container that maven initializes during the test-run, it has to be properly synchronized for dependencies using latches and static modifiers for dependencies.

markoer replied:
I tend to agree. You might be lucky and dependency would be injected, or you might not and test would finish before it began.

MMaiero committed:
interfaceName, canId and isMaster are now volatile.

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>

cdealti commented:
Let's comment out the test module for the time being.

MMaiero committed:
Commented can.test

Signed-off-by: MMaiero <matteo.maiero@eurotech.com>
